### PR TITLE
[REVIEW] Port concat.pyx to libcudf++ APIs

### PR DESCRIPTION
**Pull request by trxcllnt**
Branched from @trevorsm7's PR here: https://github.com/rapidsai/cudf/pull/4224.

Creates cython bindings to new libcudf++ concatenate APIs.

Closes https://github.com/rapidsai/cudf/issues/3926.

Edit: Waiting on https://github.com/rapidsai/cudf/pull/4346 to see if that fixes the dask_cudf test failures.

### CODE REVIEW

File: python/cudf/cudf/_libxx/concat.pyx [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    cdef vector[column_view] views
+    views.reserve(len(columns))
+    [views.push_back((<Column> col).view()) for col in columns]
+    return views
+
+
+cdef vector[table_view] make_table_views(tables):
+    cdef vector[table_view] views
+    views.reserve(len(tables))
+    [views.push_back((<Table> tbl).data_view()) for tbl in tables]
```
[the remark this answers is not in the document]
Comment (trxcllnt) [reply python/cudf/cudf/_libxx/concat.pyx]: Looking at the generated code it doesn't seem to do anything wrong. The list comprehension is treated as a new C++ scope, and the `void` return value is handled properly:
```c++
    /* ... */
      __pyx_t_6 = ((struct __pyx_vtabstruct_4cudf_6_libxx_6column_Column *)((struct __pyx_obj_4cudf_6_libxx_6column_Column *)__pyx_7genexpr__pyx_v_col)->__pyx_vtab)->view(((struct __pyx_obj_4cudf_6_libxx_6column_Column *)__pyx_7genexpr__pyx_v_col)); if (unlikely(PyErr_Occurred())) __PYX_ERR(0, 55, __pyx_L5_error)
      try {
        __pyx_v_views.push_back(__pyx_t_6);
      } catch(...) {
        __Pyx_CppExn2PyErr();
        __PYX_ERR(0, 55, __pyx_L5_error)
      }
    /* ... */
```

[the remark this answers is not in the document]
Comment (kkraus14) [reply python/cudf/cudf/_libxx/concat.pyx]: Does writing it as a more standard loop generate worse CPP? This just reads very awkwardly/confusingly from a Python perspective.

**Comment by trxcllnt** [comment]
> So instead of plumbing everything down into Column, I think it would be better to refactor concat to always go through `Frame` and call into the `Table` concat. We can keep all of the logic in a column function for type handling and whatnot.

Yeah, I'm working on updating `DataFrame._concat` to call the new `concat_tables()` method instead of concatenating each column individually.

File: python/cudf/cudf/core/column/column.py [hunk 2]
```diff
@@ -880,9 +857,7 @@ def column_empty(row_count, dtype="object", masked=False):
         data = None
         children = (
             build_column(
-                data=Buffer.empty(
-                    (row_count + 1) * np.dtype("int32").itemsize
-                ),
+                data=Buffer(cupy.zeros(row_count + 1, dtype="int32")),
```
[the remark this answers is not in the document]
Comment (trxcllnt) [reply python/cudf/cudf/core/column/column.py]: This is zero-ing out the string column's offsets buffer. This fixes the issue we were seeing with string columns thinking the strings byte counts were too large.

File: python/cudf/cudf/_libxx/cpp/concatenate.pxd [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+    cdef device_buffer concatenate_masks "cudf::concatenate_masks"(
+        const vector[column_view] columns
+    ) except +
+    cdef unique_ptr[column] concatenate_columns "cudf::concatenate"(
+        const vector[column_view] columns
+    ) except +
+    cdef unique_ptr[table] concatenate_tables \
+        "cudf::experimental::concatenate"(
+        const vector[table_view] tables
+    ) except +
```
[the remark this answers is not in the document]
Comment (trxcllnt) [reply python/cudf/cudf/_libxx/cpp/concatenate.pxd]: @trevorsm7 thoughts?

File: python/cudf/cudf/core/frame.py [hunk 4]
```diff
@@ -34,6 +37,135 @@ class Frame(libcudfxx.table.Table):
...
+        from cudf.core.column.column import build_categorical_column
+
+        if ignore_index:
+            index = cudf.core.index.RangeIndex(sum(map(len, objs)))
+        elif isinstance(objs[0].index, cudf.core.multiindex.MultiIndex):
+            index = cudf.core.multiindex.MultiIndex._concat(
+                [o.index for o in objs]
+            )
+        else:
+            index = cudf.core.index.Index._concat([o.index for o in objs])
```
[the remark this answers is not in the document]
Comment (trxcllnt) [reply python/cudf/cudf/core/frame.py]: I left this in here because I'm not sure whether/how `concatenate` should handle MultiIndex at the cython/libcudf layer. Is libcudf expected to handle MultiIndexes? Should this branching happen in the Cython instead and use `table.view()` if normal index, `tbl.data_view()` if MultiIndex?

[the remark this answers is not in the document]
Comment (kkraus14) [reply python/cudf/cudf/core/frame.py]: @trxcllnt if you call `.view()` it just has all of the index columns (whether Index or MultiIndex) as the first columns of the table. Then when passing in `index_names` in the `from_unique_ptr` call it reads those from the left most columns first. You should be able to just call `.view()` and pass the index names appropriate and everything will just work in a single libcudf API call.

File: python/cudf/cudf/core/column/column.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (kkraus14) [reply python/cudf/cudf/core/column/column.py]: I'm guessing the only time this function is called is specifically when we need to return a zero sized `StringColumn` for Pandas compatibility so I guess this makes sense.

File: python/cudf/cudf/core/frame.py [hunk 5]
```diff
@@ -42,129 +43,183 @@ def _concat(cls, objs, axis=0, ignore_index=False):
...
+                        cols[idx] = cols[idx].astype(dtype)
+
+        def reassign_categories(categories, cols, col_idxs):
+            for name, idx in zip(cols, col_idxs):
+                if idx in categories:
+                    cols[name] = build_categorical_column(
+                        categories=categories[idx],
+                        codes=cols[name],
+                        mask=cols[name].mask,
+                    )
```
Comment (kkraus14) [inline python/cudf/cudf/core/frame.py]: In case we have sliced inputs for whatever reason:
```suggestion
                    cols[name] = build_categorical_column(
                        categories=categories[idx],
                        codes=cols[name],
                        mask=cols[name].base_mask,
                        offset=cols[name].offset,
                        size=cols[name].size
                    )
```

File: python/cudf/cudf/core/frame.py [hunk 6]
```diff
@@ -42,129 +43,183 @@ def _concat(cls, objs, axis=0, ignore_index=False):
...
+        for i, cols in enumerate(columns):
+            table_cols = cols[first_data_column_position:]
+            table_names = indices[first_data_column_position:]
+            table = cls(data=dict(zip(table_names, table_cols)))
+            if 1 == first_data_column_position:
+                table._index = as_index(cols[0])
+            elif first_data_column_position > 1:
+                index_cols = cols[:first_data_column_position]
+                index_names = indices[:first_data_column_position]
+                table._index = cls(data=dict(zip(index_names, index_cols)))
```
Comment (kkraus14) [inline python/cudf/cudf/core/frame.py]: `cls` would be `DataFrame` here for example where I think you want to construct a `MultiIndex` instead of a `DataFrame`

File: python/cudf/cudf/core/frame.py [hunk 7]
```diff
@@ -42,129 +43,183 @@ def _concat(cls, objs, axis=0, ignore_index=False):
...
-                categories=cats,
-                codes=out[name]._column,
-                mask=out[name]._column.mask,
-            )
+            if first_data_column_position <= 1:
+                # If only one index column, reassign its name
+                out._index.name = objs[0]._index.name
+            else:
+                # If more than one index column, reassign MultiIndex names
+                out._index.names = objs[0]._index.names
```
Comment (kkraus14) [inline python/cudf/cudf/core/frame.py]: Can always just do `out._index.names = objs[0]._index.names`

File: python/cudf/cudf/core/frame.py [hunk 8]
```diff
@@ -42,129 +43,183 @@ def _concat(cls, objs, axis=0, ignore_index=False):
...
+            out.columns = names
 
         # TODO (ptaylor):
         # libcudf.nvtx.nvtx_range_pop()
 
-        # Return the columns in the desired out order since this might've
-        # changed if we reassigned categorical categories
-        return out[list(unique_columns_ordered_ls)]
+        return out[list(names)]
```
Comment (kkraus14) [inline python/cudf/cudf/core/frame.py]: What is in `out` here that we want to select out?

Comment (trxcllnt) [reply python/cudf/cudf/core/frame.py]: `out` is the concatenated DF -- this is selecting columns in the order of `names` in case the columns aren't in that order already.

File: python/cudf/cudf/core/frame.py (hunk 6 above)
Comment (trxcllnt) [reply python/cudf/cudf/core/frame.py]: I did try to construct a MultiIndex first, but after checking `table.pyx` logic, that seemed unnecessary. These are temporary Tables constructed for the libcudf `concatenate` call below, and `table.pyx` indicates all it needs is a Table-like value as its `_index` in order to create the combined list of column_views.

Comment (kkraus14) [reply python/cudf/cudf/core/frame.py]: Do we have a test concatenating DataFrames with MultiIndexes? If not can we add one?

Comment (kkraus14) [reply python/cudf/cudf/core/frame.py]: As pointed out by @trxcllnt offline:
https://github.com/rapidsai/cudf/blob/branch-0.13/python/cudf/cudf/tests/test_concat.py#L153
https://github.com/rapidsai/cudf/blob/branch-0.13/python/cudf/cudf/tests/test_dataframe.py#L955-L998

File: cpp/tests/io/orc_test.cu [hunk 9]
```diff
@@ -20,11 +20,13 @@
 #include <tests/utilities/cudf_gtest.hpp>
 #include <tests/utilities/type_lists.hpp>
 
+#include <cudf/concatenate.hpp>
 #include <cudf/io/functions.hpp>
 #include <cudf/strings/string_view.cuh>
 #include <cudf/strings/strings_column_view.hpp>
 #include <cudf/table/table.hpp>
 #include <cudf/table/table_view.hpp>
+#include <cudf/concatenate.hpp>
```
[the remark this answers is not in the document]
Comment (trevorsm7) [reply cpp/tests/io/orc_test.cu]: Hm, my PR only has the 2nd include at line 28/29. Can you remove the extra one?

File: python/cudf/cudf/core/frame.py [hunk 10]
```diff
@@ -213,6 +213,8 @@ def reassign_categories(categories, cols, col_idxs):
         # Reassign index and column names
         if isinstance(objs[0].columns, pd.MultiIndex):
             out.columns = objs[0].columns
+            out._index.name = objs[0]._index.name
+            out._index.names = objs[0]._index.names
         else:
             out.columns = names
             out._index.name = objs[0]._index.name
```
Comment (kkraus14) [inline python/cudf/cudf/core/frame.py]: Can move both of these outside of the branching, no?

File: python/cudf/cudf/core/multiindex.py [hunk 11]
```diff
@@ -123,7 +123,7 @@ def names(self):
     def names(self, value):
         value = [None] * self.nlevels if value is None else value
         assert len(value) == self.nlevels
-        self._names = value
+        self._names = [n for n in value]
```
Comment (kkraus14) [inline python/cudf/cudf/core/multiindex.py]: I think it's a bad assumption to assume `value` here can be efficiently iterated over. Maybe just call `list(value)`?

File: cpp/tests/io/orc_test.cu [hunk 12]
```diff
@@ -20,12 +20,12 @@
 #include <tests/utilities/cudf_gtest.hpp>
 #include <tests/utilities/type_lists.hpp>
 
+#include <cudf/concatenate.hpp>
```
Comment (kkraus14) [inline cpp/tests/io/orc_test.cu]: Is this change needed here?
